csrc/transformer/inference/includes/inference_context.h

```diff
@@ -97,12 +97,13 @@ class Context {
         if (!_free_memory_size) { cudaMemGetInfo(&_free_memory_size, &total_size); }
 
         size_t activation_size = 16 * hidden_dim * batch_size;
+        size_t temp_size = 2 * hidden_dim * batch_size;
```

Comment:
Can we change this to max_prompt_length * max_prompt_length  * batch_size * heads, which is the memory required for doing the attention part? We can either define a macro for max_prompt_length  and set it to a large-enough input-length (like 2K) or we can get it from the input? I would prefer the former.

Reply:
In such case GenWorkSpace requires num_heads, that is not provided by caller. It's easy to pass current prompt_length, but num_heads requires changing signature (args list) for `ds_qkv_gemm`, and `ds_linear_layer`, maybe it's better to make additional call for pt for internal memory allocation? With configuration, like prompt_length, max_sequence_length for model (2048 for GPT-neo for example)?

Reply:
Got it, that makes sense. By the way, I think the max_sequence_length is not always limited to a number for all models. For example, in BLOOM architecture you can potentially produce as many tokens as needed. So, let's not get it from a config and we can just assume a large enough number like 8K (if the available memory allows).

Reply:
I'll try to implement workspace allocation. And I think, it's better to write in docs, that DeepSpeed inference doesn't support parallel (in multiple threads) or interleaved generation (few tokens for one prompt, that few tokens for another, and few from first one), because internal state will be corrupted

Reply:
I agree, but let's first finish this part as it is an urgent issue, and we can get to the other issue as you suggested. We can discuss it in another thread/PR.
Thanks,
Reza

Reply:
Please let me know if you need help to finish this PR asap.

Reply:
Memory allocation updated.